### PR TITLE
Colorize Version.Requirement source in Inspect protocol

### PR DESCRIPTION
**Pull request by eksperimental**
This is to avoid ambiguity in cases like

    iex> Version.parse_requirement!("~> 1.2.3 or < 3.4.5 and > 5.6.7")
    #Version.Requirement<~> 1.2.3 or < 3.4.5 and > 5.6.7>

Before:
image

After:
image

### CODE REVIEW

**Comment by eksperimental** [comment]
Initially I tried to escape, not to colorized, but I did not like the idea.
Now i realized that when interpolating the output it is still ambiguous,

```elixir
iex(7)> IO.puts "requirement: #{inspect(vr)}"   
requirement: #Version.Requirement<~> 1.2.3 or < 3.4.5 and > 5.6.7>
```


Maybe we should escape and colorize?

**Comment by michallepicki** [comment]
@eksperimental Orange text background on the Before screenshot is your text selection, correct? And you mean that the user could easily miss the first `~>`?

**Comment by eksperimental** [comment]
> @eksperimental Orange text background on the Before screenshot is your text selection, correct? And you mean that the user could easily miss the first `~>`?

Sorry, i know my description was not clear enough. What is ambiguous is the use of  `>` in the source of the version requirement, which closes the opening `<` in  `#Version.Requirement<`

**Comment by josevalim** [comment]
@eksperimental what if we put it between quotes and escape it as a string?

    #Version.Requirement<"~> 2.0">

The string is more likely the shape we will see the requirement and it aligns with how we render some other stuff like `MapSet`.

**Comment by eksperimental** [comment]
> @eksperimental what if we put it between quotes and escape it as a string?
> 
> ```
> #Version.Requirement<"~> 2.0">
> ```
> 
> The string is more likely the shape we will see the requirement and it aligns with how we render some other stuff like `MapSet`.

Yes! I guess that is the way to got. In fact t:Version.requirement/0  is `String.t()` | `Version.Requirement.t()`

**Comment by wojtekmach** [comment]
would brackets help?

`#Version.Requirement[~> 2.0]`

we use those for calendar structs.

**Comment by eksperimental** [comment]
> would brackets help?
> 
> `#Version.Requirement[~> 2.0]`
> 
> we use those for calendar structs.

So which one is clearer when doing string interpolation?

square brackets:
> requirement: #Version.Requirement[~> 1.2.3 or < 3.4.5 and > 5.6.7]

double quotes:
> requirement: #Version.Requirement<"~> 1.2.3 or < 3.4.5 and > 5.6.7">

**Comment by josevalim** [comment]
Things like doctests assume `#...<`, so we should stick with `<`.

**Comment by eksperimental** [comment]
This is what I've got so far:

image

I am not escaping the double-quotes, because they are not allowed in versions, and I don't see how they will ever have a double quote, but let me know if you think there is an edge case

File: lib/elixir/lib/version.ex [hunk 1]
```diff
@@ -677,7 +677,13 @@ defimpl String.Chars, for: Version.Requirement do
...
-  def inspect(%Version.Requirement{source: source}, _opts) do
-    "#Version.Requirement<" <> source <> ">"
+  def inspect(%Version.Requirement{source: source}, opts) do
+    formatted =
+      ("\"" <> source <> "\"")
+      |> Inspect.Algebra.color(:string, opts)
+      |> Inspect.Algebra.format(:infinity)
+      |> IO.iodata_to_binary()
+
+    "#Version.Requirement<" <> formatted <> ">"
```
Comment (josevalim) [inline lib/elixir/lib/version.ex:687]: You should not render inside inspect. You want more something like this (untested):

```suggestion
    inner = Inspect.Algebra.color("\"" <> source <> "\"", :string, opts)
    Inspect.Algebra.concat(["#Version.Requirement<", inner, ">"])
```

Comment (eksperimental) [reply lib/elixir/lib/version.ex]: Got you. Fixed

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by eksperimental** [comment]
Thank you everyone for chiming in.

**Pull request merged**